### PR TITLE
Add create and write_multiscale() usage to README

### PR DESCRIPTION
**Pull request by will-moore**
When compared to other OME-NGFF writing tools
(see https://forum.image.sc/t/ndtiff-to-ome-zarr-ngff/58856/2) I notice that we don't provide good docs on how to write OME-NGFF.

This adds code example to the README.
It also adds the `ome_zarr create` command to the README.

To test, try running the code and opening in `napari` as described.
Should get this:

Screenshot 2021-10-21 at 09 54 33

### CODE REVIEW

**Comment by joshmoore** [comment]
Would it help if we define the "API usage" doc location? (One option would be doctests)

**Comment by will-moore** [comment]
@joshmoore Yes, that would be great. I don't know if doctests would work for a sample code of this size that uses multiple API calls (which don't have testable return types)?

**Comment by joshmoore** [comment]
Ok. Let's talk options:
 * markdown in this repo
 * github pages
 * read the docs

I assume we'll eventually want RTD so maybe it's time to just go ahead and do it?

**Comment by will-moore** [comment]
Added another method that was useful to user at https://github.com/ome/napari-ome-zarr/issues/23#issuecomment-990853332
Also shortened the code sample to make it less painful to have in the README, which I still like.
But if it's still too long I could move it to another md file. I just don't expect to have time to setup readthedocs anytime soon.

File: README.rst [hunk 1]
```diff
@@ -61,6 +61,85 @@ To download all the resolutions and metadata for an image::
...
+create
+======
+
+To create sample OME-Zarr image from the `skimage coins <https://scikit-image.org/docs/stable/api/skimage.data.html#skimage.data.coins>`_
+image::
+
+    # Create an OME-Zarr image in coinsdata/ dir:
+    $ ome_zarr create coinsdata
+
+writing OME-Zarr
```
[the remark this answers is not in the document]
Comment (joshmoore) [reply README.rst:76]: :+1:

**Comment by sbesson** [comment]
@will-moore imported the repository configuration into readthedocs and activated the `Build pull requests for this project` setting so that we can have the documentation hosted for review - see status checks and https://ome-zarr-py--121.org.readthedocs.build/en/121/.

File: .readthedocs.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#    configuration: docs/conf.py
+
+# If using Sphinx, optionally build your docs in additional formats such as PDF
+# formats:
+#    - pdf
+
+# Optionally declare the Python requirements required to build your docs
+python:
+  install:
+    - requirements: requirements.txt
```
Comment (sbesson) [inline .readthedocs.yml:33]: Reading https://docs.readthedocs.io/en/stable/config-file/v2.html#python-install, it should be possible to install  `ome_zarr`  and its dependencies without creating a new requirements txt file using the following syntax:

```
python:
  install:
    - method: pip
      path: .
```

File: .readthedocs.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+submodules:
+  include: all
```
Comment (sbesson) [inline .readthedocs.yml:36]: Is this necessary?

File: docs/source/api.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+API
+===
+
+.. autofunction:: ome_zarr.writer.write_image
```
Comment (sbesson) [inline docs/source/api.rst:4]: Potentially for a follow-up https://github.com/ome/omero-rois/blob/master/docs/index.rst uses `automodule` which would have the benefit of avoiding having to include every API by hand

File: docs/source/conf.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+sys.path.insert(0, pathlib.Path(__file__).parents[2].resolve().as_posix())
+
+
+extensions = [
+    "sphinx.ext.doctest",
+    "sphinx.ext.autodoc",
+]
+
+# use index.rst instead of contents.rst
+master_doc = "index"
```
Comment (sbesson) [inline docs/source/conf.py:14]: Also as a follow-up, I assume extra standard Sphinx metadata (project, copyright) can be added to this document

**Review by sbesson — COMMENTED** [review]
Overall, very enthusiastic about the rendering https://ome-zarr-py--121.org.readthedocs.build/en/121/. Personally, I think that's a very scalable way forward to structure and expand the reference documentation about this library.

Although the titles could be reviewed, I like the split into the 4 top-level sections (intro, cli, library, api) and I suspect a similar structure could be used in other repos e.g. `omero-metadata`.

If we decide to go  down this route, my vote for reducing the content in the README to the minimal viable version (e.g. installation + cli example + license) and point to the readthedocs as the reference documentation.

In the spirit of "The best is the enemy of the good", I strongly argue for not trying to fix everything, especially the content, as part of this PR. Assuming we collectively agree on using this infrastructure for the ref documetnation, my vote would be to limit this PR to introducing the backbone, get the docs published under readthedocs and capture next todos as issues that can be addressed separately

In terms of infrastructure, my biggest concern is around the migration of `install_requires` into `requirements.txt` which is possibly unnecessary. Also we might need to agree on the namespace of the documentation `ome-zarr.readthedocs` vs `ome-zarr-py.readthedocs`

**Comment by will-moore** [comment]
Thanks @sbesson. I think those commits should address your points above.

**Review by sbesson — APPROVED** [review]
Thanks,

From my side the only question left is whether the new section (which motivated this PR) explaining how to use the API to create an image should be introduced in the README at all or whether this will be cleaned up eventually with readthedocs becoming the reference point. In the latter case, we should likely not introduce it in this first place to save time and add appropriate links in a second step./

Otherwise, the infrastructure generating the documentation all looks fine to me. I'll leave the other mandated reviewers to confirm they are happy with this approach.

**Comment by will-moore** [comment]
I can remove all the stuff from the README that is covered by the docs, but I should replace it with a link to the docs? What will that link be??
Follow-up TODO: docs versioning

File: README.rst [hunk 6]
```diff
@@ -115,6 +164,16 @@ specifying ``major``, ``minor`` or ``patch`` depending on whether the developmen
 Remember to ``git push`` all commits and tags.
 
 
+Documentation
+-------------
+
+Documentation will be automatically build with `readthedocs`.
```
Comment (jburel) [inline README.rst:170]:
```suggestion
Documentation will be automatically built with `readthedocs`.
```

File: docs/source/cli.rst [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+You also need to specify which columns from the CSV to use, e.g. `"area,X,Y,Width,Height"`.
+You can also specify the column types (as in https://github.com/ome/omero-metadata/)
+to specify the data-type for each column (string by default).
+
+ - `d`: `DoubleColumn`, for floating point numbers
+ - `l`: `LongColumn`, for integer numbers
+ - `s`: `StringColumn`, for text
+ - `b`: `BoolColumn`, for true/false
+
+Use e.g. `#d` as a suffix in the column name to denote a `float` column, no spaces etc:
```
Comment (jburel) [inline docs/source/cli.rst:70]:
```suggestion
Use e.g. `#d` as a suffix in the column name to denote a `float` column, no spaces etc.:
```

File: docs/source/index.rst [hunk 8]
```diff
@@ -0,0 +1,15 @@
+Welcome to ome-zarr-py documentation!
```
Comment (jburel) [inline docs/source/index.rst:1]:
```suggestion
Welcome to ome-zarr-py documentation
```

File: docs/source/index.rst [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+=====================================
```
Comment (jburel) [inline docs/source/index.rst:2]:
```suggestion
====================================
```

**Comment by jburel** [comment]
Some inconsistencies in the underline used for section depending on the pages.

File: docs/source/cli.rst [hunk 10]
```diff
@@ -0,0 +1,79 @@
+
+Command-line
```
Comment (pwalczysko) [inline docs/source/cli.rst:2]: I guess this title is fed into the main doc of the Readthedocs page https://ome-zarr-py--121.org.readthedocs.build/en/121/intro.html ?
Both here, but mainly in https://ome-zarr-py--121.org.readthedocs.build/en/121/intro.html it needs a bit more wording imho, such as 
``Command-Line interface`` or ``Command-line examples`` or similar. Command-line by its own is very terse...

File: docs/source/intro.rst [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+===========
+
+Experimental support for multi-resolution images stored in Zarr filesets, according to the `OME zarr spec`_.
+
+
+Features
+--------
+
+- Use as a image reader plugin for `napari`_. See `napari-ome-zarr`_.
+- Simple :doc:`cli` to read and download conforming OME-NGFF filesets.
```
Comment (pwalczysko) [inline docs/source/intro.rst:12]: see comment about the ``Command-line`` [above](https://github.com/ome/ome-zarr-py/pull/121/files#r824742227)

File: docs/source/intro.rst [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ome-zarr-py
+===========
+
+Experimental support for multi-resolution images stored in Zarr filesets, according to the `OME zarr spec`_.
+
+
+Features
+--------
+
+- Use as a image reader plugin for `napari`_. See `napari-ome-zarr`_.
```
Comment (pwalczysko) [inline docs/source/intro.rst:11]: When I went to napari-ome-zarr as instructed, there is no mention of ome-zarr-py there. Confusing ?

Comment (pwalczysko) [reply docs/source/intro.rst:11]:
```suggestion
- Use as an image reader plugin for `napari`_. See `napari-ome-zarr`_.
```

fix typo

File: docs/source/intro.rst [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+- Simple :doc:`cli` to read and download conforming OME-NGFF filesets.
+- :doc:`python` for reading and writing OME-NGFF filesets.
```
Comment (pwalczysko) [inline docs/source/intro.rst:13]: These three very important bullet points start each time differently - once with a verb, once with "Simple..." (there is no verb in that sentence) and once even with a link.
Understood this is automatically generated, but it does not read very well.

File: docs/source/intro.rst [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+Installation
+------------
+
+Install the latest release of `ome-zarr`_ from PyPI::
+
+    pip install ome-zarr
+
+
+Install developer mode to run from your current branch::
```
Comment (pwalczysko) [inline docs/source/intro.rst:24]: Maybe better 

``Installation for developers`` ?

File: docs/source/intro.rst [hunk 15]
```diff
@@ -0,0 +1,44 @@
+===========
+ome-zarr-py
+===========
+
+Experimental support for multi-resolution images stored in Zarr filesets, according to the `OME zarr spec`_.
```
Comment (pwalczysko) [inline docs/source/intro.rst:5]: OME zarr spec link leads to a repo called ome/ngff - potential for confusion ? (no mention of ``zarr`` on that page)

File: docs/source/python.rst [hunk 16]
```diff
@@ -0,0 +1,80 @@
+ome_zarr python library
+=======================
+
+writing OME-NGFF images
+-----------------------
+
+The principle entry-point for writing OME-NGFF images is :py:func:`ome_zarr.writer.write_image`.
+This takes an n-dimensional `numpy` array and writes it to the specified `zarr group` according
+to the OME-NGFF specification.
```
Comment (pwalczysko) [inline docs/source/python.rst:9]: Here it is called OME-NGFF specification. On the top of the main readthedocs page https://ome-zarr-py--121.org.readthedocs.build/en/121/intro.html#license it is called ome-zarr-spec.  Possibly this naming could be unified within one single doc here ?

**Review by pwalczysko — CHANGES_REQUESTED** [review]

File: docs/source/cli.rst (hunk 10 above)
Comment (will-moore) [reply docs/source/cli.rst]: Use 'Command line Tools'

**Comment by will-moore** [comment]
@pwalczysko Thanks for the comments and fixes. Hopefully mostly addressed by those commits.

File: docs/source/python.rst [hunk 17]
```diff
@@ -0,0 +1,80 @@
+ome_zarr python library
+=======================
+
+writing OME-NGFF images
```
Comment (pwalczysko) [inline docs/source/python.rst:4]:
```suggestion
Writing OME-NGFF images
```

Comment (pwalczysko) [reply docs/source/python.rst:4]: More important comment though:

The bullet point with a link to this doc is tallking about writing ``filesets``. Should this title and text also rather talk about filesets ?

**Comment by pwalczysko** [comment]
Thank you @will-moore , sorry for not spotting the new comment https://github.com/ome/ome-zarr-py/pull/121/files#r824817468 earlier - it became obvious only after you cleaned the bullet points so nicely.

One more comment https://github.com/ome/ome-zarr-py/pull/121/files#r824817468 and typo correction (depends on the comment https://github.com/ome/ome-zarr-py/pull/121/files#r824817468), otherwise LGTM

File: docs/source/python.rst (hunk 16 above)
Comment (will-moore) [reply docs/source/python.rst:9]: I've updated `ome-zarr-spec` to OME NGFF spec.

**Comment by sbesson** [comment]
Note that as readthedocs has been activated, the checks will start to fail for new PRs without these infrastructure changes included.

@joshmoore @pwalczysko @jburel any conceptual blockers to moving forward with this solution? Any blocker you would like to see addressed before merging?

**Review by pwalczysko — APPROVED** [review]

**Comment by joshmoore** [comment]
> As noted above, the two most obvious options from my side are https://ome-zarr-py.readthedocs.io and https://ome-zarr.readthedocs.io with the former one being configured at the moment.

Since zarr-python also uses zarr.rtd.io I could see doing the same here (under the assumption that ome-zarr-java if/when it exists would live under https://javadocs.io/, e.g.)

**Comment by sbesson** [comment]
> Since zarr-python also uses zarr.rtd.io I could see doing the same here (under the assumption that ome-zarr-java if/when it exists would live under https://javadocs.io/, e.g.)

:+1: I also like find the idea of mirroring the behavior of zarr-python attractive. The only caveat I would raise though is that https://zarr.readthedocs.io also includes the Zarr specification which legitimates the choice of dropping `-python` in that case. But this would not apply here

**Comment by joshmoore** [comment]
Agreed for v2. But https://zarr-specs.readthedocs.io/ is ready and waiting :)

**Comment by sbesson** [comment]
Deleted and recreated the readthedocs project using ome-zarr as the name - see https://ome-zarr--121.org.readthedocs.build/en/121/.

**Comment by joshmoore** [comment]
> any conceptual blockers to moving forward with this solution? Any blocker you would like to see addressed before merging?

Based on #175's failure, I'd say go for it (or less ideally, we need to deactivate RTD)

**Comment by sbesson** [comment]
Now published as https://ome-zarr.readthedocs.io

**Pull request merged**